File: apps/minecraft/spigot/src/main/java/crabcraft/net/crabUtilities/update/UpdateDownloader.java
package crabcraft.net.crabUtilities.update;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.file.AtomicMoveNotSupportedException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardCopyOption;
import java.nio.file.StandardOpenOption;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.time.Duration;

public class UpdateDownloader {

    private final HttpClient http;
    private final String token;
    private final String userAgent;

    public UpdateDownloader(String token, String userAgent) {
        this.http = HttpClient.newBuilder()
                .connectTimeout(Duration.ofSeconds(10))
                .followRedirects(HttpClient.Redirect.NORMAL)
                .build();
        this.token = token;
        this.userAgent = userAgent;
    }

    public Path download(ReleaseInfo info, Path targetDir, String targetFilename)
            throws IOException, InterruptedException {

        Files.createDirectories(targetDir);
        Path target = targetDir.resolve(targetFilename);
        Path part = targetDir.resolve("." + targetFilename + ".part");

        String expected = null;
        if (info.checksumUrl() != null) {
            expected = fetchChecksum(info.checksumUrl(), info.jarAssetName());
        }

        MessageDigest sha;
        try {
            sha = MessageDigest.getInstance("SHA-256");
        } catch (NoSuchAlgorithmException e) {
            throw new IOException("SHA-256 not available", e);
        }

        try {
            HttpRequest.Builder b = HttpRequest.newBuilder()
                    .uri(URI.create(info.jarUrl()))
                    .timeout(Duration.ofMinutes(5))
                    .header("Accept", "application/octet-stream")
                    .header("User-Agent", userAgent)
                    .GET();
            if (token != null && !token.isEmpty()) {
                b.header("Authorization", "Bearer " + token);
            }
            HttpResponse<InputStream> res = http.send(b.build(), HttpResponse.BodyHandlers.ofInputStream());
            if (res.statusCode() / 100 != 2) {
                throw new IOException("HTTP " + res.statusCode() + " downloading " + info.jarUrl());
            }
            try (InputStream in = res.body();
                 OutputStream out = Files.newOutputStream(part,
                         StandardOpenOption.CREATE, StandardOpenOption.TRUNCATE_EXISTING)) {
                byte[] buf = new byte[8192];
                int n;
                while ((n = in.read(buf)) > 0) {
                    sha.update(buf, 0, n);
                    out.write(buf, 0, n);
                }
            }

            String actual = toHex(sha.digest());
            if (expected != null && !expected.equalsIgnoreCase(actual)) {
                throw new UpdateExceptions.ChecksumMismatchException(
                        "SHA-256 mismatch for " + info.jarAssetName()
                                + ": expected " + expected + ", got " + actual);
            }

            try {
                Files.move(part, target,
                        StandardCopyOption.ATOMIC_MOVE, StandardCopyOption.REPLACE_EXISTING);
            } catch (AtomicMoveNotSupportedException e) {
                Files.move(part, target, StandardCopyOption.REPLACE_EXISTING);
            }
            return target;
        } finally {
            try {
                Files.deleteIfExists(part);
            } catch (IOException ignored) {
            }
        }
    }

    private String fetchChecksum(String url, String assetName) throws IOException, InterruptedException {
        HttpRequest.Builder b = HttpRequest.newBuilder()
                .uri(URI.create(url))
                .timeout(Duration.ofSeconds(30))
                .header("Accept", "text/plain")
                .header("User-Agent", userAgent)
                .GET();
        if (token != null && !token.isEmpty()) {
            b.header("Authorization", "Bearer " + token);
        }
        HttpResponse<String> res = http.send(b.build(), HttpResponse.BodyHandlers.ofString());
        if (res.statusCode() / 100 != 2) {
            return null;
        }
        for (String line : res.body().split("\\R")) {
            String trimmed = line.trim();
            if (trimmed.isEmpty() || trimmed.startsWith("#")) continue;
            String[] parts = trimmed.split("\\s+", 2);
            if (parts.length != 2) continue;
            String name = parts[1].startsWith("*") ? parts[1].substring(1) : parts[1];
            if (name.equalsIgnoreCase(assetName)) {
                return parts[0];
            }
        }
        return null;
    }

    private static String toHex(byte[] bytes) {
        StringBuilder sb = new StringBuilder(bytes.length * 2);
        for (byte b : bytes) sb.append(String.format("%02x", b));
        return sb.toString();
    }
}

File: apps/minecraft/shared/src/main/kotlin/crabcraft/net/crabUtilities/update/UpdateDownloader.kt
package crabcraft.net.crabUtilities.update

import java.io.IOException
import java.net.URI
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse
import java.nio.file.AtomicMoveNotSupportedException
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.StandardCopyOption
import java.nio.file.StandardOpenOption
import java.security.MessageDigest
import java.security.NoSuchAlgorithmException
import java.time.Duration
import java.util.HexFormat
import java.util.regex.Pattern

open class UpdateDownloader(private val token: String?, private val userAgent: String) {
    private val http =
        HttpClient.newBuilder()
            .connectTimeout(Duration.ofSeconds(10))
            .followRedirects(HttpClient.Redirect.NORMAL)
            .build()

    @Throws(IOException::class, InterruptedException::class)
    open fun download(info: ReleaseInfo, targetDir: Path, targetFilename: String): Path {
        Files.createDirectories(targetDir)
        val target = targetDir.resolve(targetFilename)
        val part = targetDir.resolve(".$targetFilename.part")
        val checksumUrl = info.checksumUrl()
        if (checksumUrl == null || checksumUrl.all { Character.isWhitespace(it) }) {
            throw IOException("Release is missing SHA256SUMS.txt")
        }
        val expected = fetchChecksum(checksumUrl, info.jarAssetName())
        if (expected == null || !expected.matches(Regex("(?i)^[0-9a-f]{64}$"))) {
            throw IOException("Release checksum is missing or invalid for ${info.jarAssetName()}")
        }
        if (info.size() <= 0 || info.size() > MAX_UPDATE_BYTES) {
            throw IOException("Release asset has an invalid size: ${info.size()}")
        }
        val sha =
            try {
                MessageDigest.getInstance("SHA-256")
            } catch (e: NoSuchAlgorithmException) {
                throw IOException("SHA-256 not available", e)
            }

        try {
            val response =
                http.send(
                    request(info.jarUrl(), Duration.ofMinutes(5)),
                    HttpResponse.BodyHandlers.ofInputStream(),
                )
            if (response.statusCode() / 100 != 2) {
                response.body().close()
                throw IOException("HTTP ${response.statusCode()} downloading ${info.jarUrl()}")
            }
            val contentLength = response.headers().firstValueAsLong("Content-Length").orElse(-1L)
            if (contentLength > info.size() || contentLength > MAX_UPDATE_BYTES) {
                response.body().close()
                throw IOException("Release asset exceeds its declared size")
            }
            response.body().use { input ->
                Files.newOutputStream(part, StandardOpenOption.CREATE, StandardOpenOption.TRUNCATE_EXISTING).use {
                    output ->
                    val buffer = ByteArray(8192)
                    var downloaded = 0L
                    while (true) {
                        val count = input.read(buffer)
                        if (count <= 0) break
                        downloaded += count
                        if (downloaded > info.size() || downloaded > MAX_UPDATE_BYTES) {
                            throw IOException("Release asset exceeds its declared size")
                        }
                        sha.update(buffer, 0, count)
                        output.write(buffer, 0, count)
                    }
                    if (downloaded != info.size()) {
                        throw IOException("Release asset size mismatch: expected ${info.size()}, got $downloaded")
                    }
                }
            }
            val actual = HexFormat.of().formatHex(sha.digest())
            if (!expected.equals(actual, ignoreCase = true)) {
                throw UpdateExceptions.ChecksumMismatchException(
                    "SHA-256 mismatch for ${info.jarAssetName()}: expected $expected, got $actual"
                )
            }
            try {
                Files.move(part, target, StandardCopyOption.ATOMIC_MOVE, StandardCopyOption.REPLACE_EXISTING)
            } catch (_: AtomicMoveNotSupportedException) {
                Files.move(part, target, StandardCopyOption.REPLACE_EXISTING)
            }
            return target
        } finally {
            try {
                Files.deleteIfExists(part)
            } catch (_: IOException) {}
        }
    }

    private fun request(url: String, timeout: Duration): HttpRequest {
        val request =
            HttpRequest.newBuilder()
                .uri(URI.create(url))
                .timeout(timeout)
                .header("Accept", "application/octet-stream")
                .header("User-Agent", userAgent)
                .GET()
        if (!token.isNullOrEmpty()) request.header("Authorization", "Bearer $token")
        return request.build()
    }

    @Throws(IOException::class, InterruptedException::class)
    private fun fetchChecksum(url: String, assetName: String): String? {
        val response = http.send(request(url, Duration.ofSeconds(30)), HttpResponse.BodyHandlers.ofInputStream())
        if (response.statusCode() / 100 != 2) {
            response.body().close()
            return null
        }
        val body =
            response.body().use { input ->
                val bytes = input.readNBytes(MAX_CHECKSUM_BYTES + 1)
                if (bytes.size > MAX_CHECKSUM_BYTES) throw IOException("Checksum file exceeds 64 KiB")
                String(bytes, Charsets.UTF_8)
            }
        for (line in Pattern.compile("\\R").split(body)) {
            val trimmed = line.trim { it <= ' ' }
            if (trimmed.isEmpty() || trimmed.startsWith("#")) continue
            val parts = Pattern.compile("\\s+").split(trimmed, 2)
            if (parts.size != 2) continue
            val name = parts[1].removePrefix("*")
            if (name.equals(assetName, ignoreCase = true)) return parts[0]
        }
        return null
    }

    companion object {
        private const val MAX_UPDATE_BYTES = 128L * 1024 * 1024
        private const val MAX_CHECKSUM_BYTES = 64 * 1024
    }
}
